vignettes/oauth.Rmd

```diff
@@ -0,0 +1,469 @@
+
+```{r httr}
+url3 <- paste0(
+  regi_url,
+  "?name=3P%20app2",
+  "&redirect_uri=", httr::oauth_callback()
+)
+reg_resp3 <- httr::GET(url3)
+reg_resp3
+regdata3 <- httr::content(reg_resp3)
+regdata3
+```
+
+Now we set the registration data on the third-party app.
+
+```{r httr-dance}
+app <- httr::oauth_app(
+  "3P app2",
+  key = regdata3$client_id[[1]],
+  secret = regdata3$client_secret[[1]],
+  redirect_uri = httr::oauth_callback()
+)
+
+endpoint <- httr::oauth_endpoint(
+  authorize = auth_url,
+  access = toke_url
+)
+```
+
+Now we can launch the token creation.
+
+```{r httr-dance2}
+token <- oauth2_httr_login(
+  httr::oauth2.0_token(endpoint, app, cache = FALSE)
+)
+```
+
+```{r refresh}
+token
+```
+
+Without the token, the query to the resource server fails:
+
+```{r}
+httr::GET(rsapp$url("/data"))
+```
+
+With the token, it is successful.
+httr also automatically refreshes the token if needed.
+
+```{r}
+httr::content(
+  httr::GET(rsapp$url("/data"), config = token),
+  as = "text"
+)
+```
+
+## Advanced topics
+
+### Applications
+
+With these apps, or only the resource server app, you can now test your code that helps users create and store an OAuth2.0 token.
+
+Like all webfakes apps, OAuth2.0 apps are extensible: you can add your own endpoints and middleware to it.
+E.g. here we add logging via `mw_log()` and a new endpoint.
+
+```{r}
+rsapp2 <- oauth2_resource_app(
+  refresh_duration = .Machine$integer.max,
+  access_duration = 10L
+)
+logfile <- tempfile("oauth-rs-", fileext = ".log")
+rsapp2$use(mw_log(stream = logfile), .first = TRUE)
+rsapp2$get("/docs", function(req, res) {
+  res$
+    set_status(200L)$
+    send("See vignette('oauth', package = 'webfakes')")
+})
+
+rsapp2_process <- new_app_process(
+  rsapp2,
+  opts = server_opts(num_threads = 3)
+)
+```
+
+If you want to customize one of the apps or both apps a lot, it might make sense to use their source code as starting point or inspiration.
+
+### Debugging
+
+See the [usual debugging advice for webfakes apps](https://r-lib.github.io/webfakes/dev/articles/how-to.html#how-can-i-debug-an-app-).
+In particular, you can add the `mw_log()` middleware to write the log of the app to a file, like we did above.
+
+The resource app has a `/locals` endpoint, that returns all data stored in the app, this includes the tokens and the refresh tokens:
+
+```{r}
+httr::content(
+  httr::GET(rsapp$url("/locals"))
+)
+```
+
+## Case study for OAuth2.0 testing
+
+Consider a package that has a function that uses OAuth2.0 to access GitHub.
+In this section we'll show how you can use webfakes to test this function.
+
+```{r}
+gh_base <- function() {
+  Sys.getenv("FAKE_GH_API_BASE", "https://api.github.com")
+}
+gh_oauth_base <- function() {
+  Sys.getenv(
+    "FAKE_GH_AUTH_BASE",
+    "https://github.com/login/oauth"
+  )
+}
+gh_repos <- function() {
+  ghapp <- httr::oauth_app(
+    "gh_repos",
+    key = Sys.getenv("GH_CLIENT_ID"),
```

Comment:
maybe add why those are environment variables (because they are the real variables of an app?)

Reply:
ah actually it's explained later, maybe move the app registration explanations to earlier?